rtwo/drivers/openstack_network.py

```diff
@@ -148,19 +148,31 @@ def disassociate_floating_ip(self, server_id):
         deleted_ip = self.neutron.delete_floatingip(floating_ip_id)
         return
 
-    def associate_floating_ip(self, server_id):
+    def associate_floating_ip(self, server_id, external_network_id=None):
         """
         Create a floating IP on the external network
         Find port of new VM
         Associate new floating IP with the port assigned to the new VM
+        If external_network_id is not specified then an external network is
+        chosen arbitrarily. This may not work if there are multiple external
+        networks.
         """
         networks = self.list_networks()
-        external_networks = [net for net in networks 
-                             if net['router:external'] == True]
-        if not external_networks:
-            raise Exception("CONFIGURATION ERROR! No external networks found!"
-                            " Cannot associate floating ip without it!"
-                            " Create a fixed IP/port first!")
+        if external_network_id:
+            for net in networks:
+                if net['id'] == external_network_id:
+                    if net['router:external'] != True:
+                        raise Exception("Provided external network ID is not a"
+                                        " valid external network")
+                    external_network = net
```

Comment:
Ideally you'd include a break here, and set external_network a default (see earlier comment). You never need to test if something is either `== True` or `== !True`, because `if not net['router:external']` is equivalent to `if ... != True` [0]. If you're going to raise a generic exception, an assertion is typically preferred.
```
for net in networks:
    if net['id'] == external_network_id
        assert net['router:external'], "Provided external network ID is not a valid external network"
        external_network = net
        break
```

[0] In a language like python you don't want to check if something is actually True, you want to check if it's truthy.